### PR TITLE
[CDAP-19659] Refactor SparkResourceFilter to be Non-Static Instance and Instantiate it using Spark classloader

### PR DESCRIPTION
**Pull request by dli357**
Context: The SparkResourceFilter has dependencies on Spark classes, notably the `org.apache.spark.streaming.StreamingContext` class. In environments in which the parent classloader does not have access to Spark classes (e.g. Kubernetes), we must create the SparkResourceFilter using the newly-created SparkClassLoader instead of directly instantiating it which will throw NoClassDefFoundError.

### CODE REVIEW

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java [hunk 1]
```diff
@@ -211,11 +222,11 @@ private ProgramRunner createSparkProgramRunner(Injector injector,
                                                  String programRunnerClassName,
                                                  ClassLoader classLoader) {
     try {
-      ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(classLoader);
+      ClassLoader oldClassLoader = setContextClassLoader(classLoader);
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: Seems like an unnecessary change? Typically we don't use static import.

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java [hunk 2]
```diff
@@ -93,12 +95,21 @@ public ClassLoader createProgramClassLoader(CConfiguration cConf, ProgramType ty
       cConf.getBoolean(SparkRuntimeUtils.SPARK_STREAMING_CHECKPOINT_REWRITE_ENABLED);
     boolean rewriteYarnClient = cConf.getBoolean(Constants.AppFabric.SPARK_YARN_CLIENT_REWRITE);
 
+    ClassLoader sparkClassLoader = null;
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: Call this `sparkRunnerClassLoader` to be consistent with another method.

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java [hunk 3]
```diff
@@ -93,12 +95,21 @@ public ClassLoader createProgramClassLoader(CConfiguration cConf, ProgramType ty
...
+      throw new RuntimeException(e);
+    }
+    // SparkResourceFilter must be instantiated using the above classloader as it has the
+    // org.apache.spark.streaming.StreamingContext class, otherwise it will cause NoClassDefFoundError at runtime
+    // because the parent classloader does not have Spark resources loaded.
+    try {
+      FilterClassLoader.Filter sparkClassLoaderFilter = (FilterClassLoader.Filter) sparkClassLoader
+        .loadClass(SparkResourceFilter.class.getName()).newInstance();
+      return new FilterClassLoader(sparkClassLoader, sparkClassLoaderFilter);
+    } catch (Exception e) {
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: Can you combine this try catch block with the previous one? The `loadClass` and `newInstance` methods throw specific exception that you can catch them to distinguish the cases.

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkResourceFilter.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+/**
+ * A class for filtering spark resources for the classloader. It is in its own class to minimize the
+ * dependencies of this class.
+ */
+public final class SparkResourceFilter implements FilterClassLoader.Filter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SparkResourceFilter.class);
+
+  final FilterClassLoader.Filter defaultFilter = FilterClassLoader.defaultFilter();
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkResourceFilter.java]: private

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkResourceFilter.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  volatile Set<ClassPath.ResourceInfo> sparkStreamingResources;
```
Comment (chtyim) [inline cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkResourceFilter.java]: private

File: cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/NoSparkClassLoaderTestRunner.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+          throw new ClassNotFoundException("Failed to read class definition for class " + name, e);
+        }
+      } else if (name.startsWith(SPARK_PACKAGE_NAME)) {
+        // Do not load Spark classes at all.
+        throw new ClassNotFoundException("Spark classes are not to be loaded by the DelegateNoSparkCLassLoader");
+      }
+      // For non-CDAP and non-Spark classes, delegate back to original classloader.
+      return delegate.loadClass(name);
+    }
+
```
Comment (chtyim) [inline cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/NoSparkClassLoaderTestRunner.java]: Remove extra empty lines

File: cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/NoSparkClassLoaderTestRunner.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    super(getTestClassWithDelegatedClassLoader(klass));
+  }
+
+  private static Class<?> getTestClassWithDelegatedClassLoader(Class<?> klass) throws ClassNotFoundException {
+    return new DelegateNoSparkClassLoader(klass.getClassLoader()).loadClass(klass.getName());
+  }
+
+  private static class DelegateNoSparkClassLoader extends ClassLoader {
+    final ClassLoader delegate;
+    private DelegateNoSparkClassLoader(ClassLoader delegate) {
```
Comment (chtyim) [inline cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/NoSparkClassLoaderTestRunner.java:46]: Add a line to separate field and method.

File: cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java [hunk 8]
```diff
@@ -16,51 +16,46 @@
...
-
+    ClassLoader sparkClassLoader = runtimeProvider.createProgramClassLoader(cConf, ProgramType.SPARK);
+    // Load any spark streaming class.
+    sparkClassLoader.loadClass("org.apache.spark.streaming.StreamingContext");
     MasterEnvironments.setMasterEnvironment(tmpMasterEnv);
+    ClassLoaders.setContextClassLoader(tmp);
   }
-
-  @Test
-  public void testSystemAppTaskInjector() throws Exception {
```
Comment (chtyim) [inline cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java:52]: Why removing this test?

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java (hunk 3 above)
Comment (dli357) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: Fixed

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java (hunk 2 above)
Comment (dli357) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: Fixed

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkResourceFilter.java (hunk 4 above)
Comment (dli357) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkResourceFilter.java]: Fixed

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkResourceFilter.java (hunk 5 above)
Comment (dli357) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkResourceFilter.java]: Fixed

File: cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/NoSparkClassLoaderTestRunner.java (hunk 7 above)
Comment (dli357) [reply cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/NoSparkClassLoaderTestRunner.java:46]: Fixed

File: cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/NoSparkClassLoaderTestRunner.java (hunk 6 above)
Comment (dli357) [reply cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/NoSparkClassLoaderTestRunner.java]: Fixed

File: cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java (hunk 8 above)
Comment (dli357) [reply cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java:52]: When ArtifactClassLoaderFactory originally used `createProgramRunner`, it needed to be able to dynamically instantiate a [Distributed]SparkProgramRunner using the existing Guice bindings; thus, this test was added to ensure that the Guice bindings are always present. The injector was also part of the `createProgramRunner` contract.

However, after refactoring the ArtifactClassLoaderFactory to remove this logic, I don't think we need these tests anymore since we no longer instantiate the entire ProgramRunner, and the injector is not used in the `createProgramClassLoader` contract.

File: cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java (hunk 1 above)
Comment (dli357) [reply cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProvider.java]: I think this was unintentional on my part; thanks for catching, fixed.

File: cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java (hunk 8 above)
Comment (chtyim) [reply cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java:52]: The problem is that the `SystemAppTask.createInjector` method is still there and in used. This method is to test that, regardless what changed inside the guice bindings for that injector.

Comment (dli357) [reply cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java:52]: `SystemAppTask.createInjector` is separately tested in [SystemAppTaskTest.java](https://github.com/cdapio/cdap/blob/2d2ae49d40d87b96a72cbeaa767561a8ae1a041c/cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/worker/SystemAppTaskTest.java). This test is specifically intended for testing that `ProgramRunner` can be properly instantiated by a SystemAppTask injector; however, the `SystemAppTask.createInjector` method is no longer used to instantiate a program runner anywhere.

Comment (chtyim) [reply cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java:52]: I see. Ok to remove. BTW, we should all the tests against the same class in the same test class if possible.

**Review by chtyim — APPROVED** [review]

File: cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java (hunk 8 above)
Comment (dli357) [reply cdap-spark-core-base/src/test/java/io/cdap/cdap/app/runtime/spark/SparkProgramRuntimeProviderTest.java:52]: Yup; main reason I had to have it in this class in the first place is because I'm not sure how to get the runtime extension classes (i.e. SparkProgramRuntimeProvider) in appfabric. However, will keep this in mind in the future. Thanks!
